ingest/importer/submission/entity_linker.py

```diff
@@ -0,0 +1,184 @@
+                'id': project.id,
+                'relationship': 'projects'
+            })
+            entity_map.add_entity(linking_process)
+
+            # link output of process
+            entity.direct_links.append({
+                'entity': linking_process.type,
+                'id': linking_process.id,
+                'relationship': 'derivedByProcesses'
+            })
+
+            # apply all protocols to the linking process
+            for linked_protocol_id in linked_protocol_ids:
+                linking_process.direct_links.append({
+                    'entity': 'protocol',
+                    'id': linked_protocol_id,
+                    'relationship': 'protocols'
+                })
+
+            # biomaterial-biomaterial
+            # file-biomaterial
+            for linked_biomaterial_id in linked_biomaterial_ids:
+                linked_biomaterial_entity = entity_map.get_entity('biomaterial', linked_biomaterial_id)
+                linked_biomaterial_entity.direct_links.append({
+                    'entity': linking_process.type,
+                    'id': linking_process.id,
+                    'relationship': 'inputToProcesses'
+                })
+
+            # file-file
+            for linked_file_id in linked_file_ids:
+                linked_file_entity = entity_map.get_entity('file', linked_file_id)
+                linked_file_entity.direct_links.append({
+                    'entity': linking_process.type,
+                    'id': linking_process.id,
+                    'relationship': 'inputToProcesses'
+                })
+
+    def link_process(self, entity_map, linked_process_id, linking_details):
+        if not linked_process_id:
+            linked_process_id = self._generate_empty_process_id()
+
+        linking_process = self.create_or_get_process(entity_map, linked_process_id, linking_details)
+
+        return linking_process
+
+    def _validate_entity_links(self, entity_map, entity):
+        links_by_entity = entity.links_by_entity
+
+        for link_entity_type, link_entity_ids in links_by_entity.items():
+            for link_entity_id in link_entity_ids:
+                if not link_entity_type == 'process':  # it is expected that no processes are defined in any tab,
+                    # these will be created later
+                    if not self._is_valid_spreadsheet_link(entity.type, link_entity_type):
+                        raise InvalidLinkInSpreadsheet(entity, link_entity_type, link_entity_id)
+                    if not entity_map.get_entity(link_entity_type, link_entity_id):
+                        raise LinkedEntityNotFound(entity, link_entity_type, link_entity_id)
+                    if not entity_map.get_entity(link_entity_type, link_entity_id):
+                        raise LinkedEntityNotFound(entity, link_entity_type, link_entity_id)
+
+                if link_entity_type == 'process' and not len(link_entity_ids) == 1:
```

Comment:
Why not if/else?

Reply:
Won't touch this file for now.

Reply:
so that there's no inner if ? i think, I wrote this a long time ago :(